Refactor Sqlalchemy queries to 2.0 styles (Part 2)

ephraimbuddy opened the pull request:
This is a continuation of the effort to refactor the queries to sqlalchemy 2.0

ephraimbuddy committed:
Refactor Sqlalchemy queries to 2.0 styles (Part 2)

This is a continuation of the effort to refactor the queries to sqlalchemy 2.0 style

ephraimbuddy commented on airflow/api_connexion/endpoints/task_instance_endpoint.py:
```diff
@@ -255,7 +251,8 @@ def get_mapped_task_instances(
         query = query.order_by(TI.map_index.asc())
 
-    task_instances = query.offset(offset).limit(limit).all()
+    task_instances = session.execute(query.offset(offset).limit(limit)).all()
```
This query is giving me a tough time. Tests are failing and I can't figure out what's wrong. Will appreciate some help cc @jedcunningham @uranusjr @dstandish

uranusjr replied:
Looks like the `total_entries` line isn’t working as expected, values that should be excluded by filters applied above are still counted. The actual result `response.json["task_instances"]` is correct, only the `total_count` number is wrong. I’ll try digging deeper…

uranusjr replied:
Oh wait this is **not** the part that’s failing! The failing tests actually point to `get_mapped_task_instance_endpoint` above. (This is `get_task_instances`.)

uranusjr replied:
I got this. Should work now.

ephraimbuddy replied:
Ok. Thanks!

uranusjr commented on airflow/api_connexion/endpoints/xcom_endpoint.py:
```diff
@@ -93,15 +93,20 @@ def get_xcom_entry(
+    if deserialize:
+        item = session.execute(query).one_or_none()
+    else:
+        item = session.scalars(query).one_or_none()
```
I think this fine actually.

uranusjr commented on airflow/api_connexion/endpoints/user_endpoint.py:
```diff
@@ -119,7 +119,7 @@ def post_user() -> APIResponse:
     if not user:
         detail = f"Failed to add user `{username}`."
-        return Unknown(detail=detail)
+        raise Unknown(detail=detail)
```
Oops, a bug found?

ephraimbuddy replied:
Yeah. Fixed yesterday. Rebasing would fix it

uranusjr commented on airflow/api_connexion/endpoints/log_endpoint.py:
```diff
@@ -77,18 +78,17 @@ def get_log(
         .join(TaskInstance.dag_run)
-        .options(joinedload("trigger"))
-        .options(joinedload("trigger.triggerer_job"))
+        .options(joinedload(TaskInstance.trigger))
```
Does this still need to join `triggerer_job`?

ephraimbuddy replied:
It was giving errors which made sense when I tried joining it

uranusjr replied:
I think we still need to do the join though, it’s needed in `read_log_stream` and not joining it would incur at least one additional SQL fetch.

[SQLAlchemy documentation says the load options can be chained](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html) so we probably need to do something like this:

```python
.options(joinedload(TaskInstance.trigger).joinedload(Trigger.triggerer_job))
```

ephraimbuddy replied:
I will check again

uranusjr committed:
Fix total_entries counting logic

The count should happen *after* we apply the filters, and is (slightly)
different from the automatic 404 case.

ephraimbuddy committed:
Update queries in models/

ephraimbuddy committed:
fixup! Update queries in models/

ephraimbuddy committed:
Add triggerer_job in query

ephraimbuddy committed:
fixup! fixup! Update queries in models/

ephraimbuddy committed:
Revert changes to dagwarning.py

ephraimbuddy committed:
Use session.scalar

uranusjr commented on airflow/models/dag.py:
```diff
@@ -2181,9 +2199,9 @@ def clear(
+            return session.scalars(tis).all()
 
-        tis = list(tis)
+        tis = list(session.scalars(tis).all())
```
```suggestion
        tis = session.scalars(tis).all()
```

uranusjr commented on airflow/models/dag.py:
```diff
@@ -2181,9 +2199,9 @@ def clear(
 
         if dry_run:
-            return tis
+            return session.scalars(tis).all()
```
Need to update the return type annotation of this function to match. But can we safely change in the first place? Calling `all()` here can be quite slow if there are a lot of task instances, which may be pretty unexpected for `dry_run=True`.

ephraimbuddy replied:
I will update the return type as without calling `all` here, the session closes and `all` called in other places would error out

ephraimbuddy replied:
Looking at it more, it seems the return type matches. Can you verify?

uranusjr committed:
Convert DagWarning to SQLAlchemy 2.x syntax

ephraimbuddy committed:
Fix dagwarning

uranusjr commented on airflow/models/pool.py:
```diff
@@ -162,19 +163,19 @@ def slots_stats(
             query = with_row_locks(query, session=session, **nowait(session))
 
-        pool_rows: Iterable[tuple[str, int]] = query.all()
+        pool_rows: Iterable[Row] = session.execute(query).all()
```
So unfortunate this is actually worse than the `tuple` typing since we can no longer express the innter types 😞

ephraimbuddy replied:
I have removed the typing. I think there's no need for it

ephraimbuddy committed:
fix typing and remove todo comment

uranusjr commented on airflow/models/dag.py:
```diff
@@ -223,31 +238,32 @@ def get_dataset_triggered_next_run_info(
+            )
+            .group_by(
+                DagScheduleDatasetReference.dag_id,
+            )
```
```suggestion
            .join(DatasetModel, DatasetModel.id == DagScheduleDatasetReference.dataset_id)
            .group_by(DagScheduleDatasetReference.dag_id)
```

uranusjr commented on airflow/models/pool.py:
```diff
@@ -162,19 +162,19 @@ def slots_stats(
             query = with_row_locks(query, session=session, **nowait(session))
 
-        pool_rows: Iterable[tuple[str, int]] = query.all()
+        pool_rows = session.execute(query).all()
```
```suggestion
        pool_rows = session.execute(query)
```

uranusjr commented on airflow/models/dagrun.py:
```diff
@@ -283,17 +286,17 @@ def active_runs_of_dags(
+            query = query.where(cls.state.in_([State.RUNNING, State.QUEUED]))
         query = query.group_by(cls.dag_id)
-        return {dag_id: count for dag_id, count in query.all()}
+        return {dag_id: count for dag_id, count in session.execute(query).all()}
```
```suggestion
        return {dag_id: count for dag_id, count in session.execute(query)}
```

uranusjr replied:
(I think this can actually be `dict(session.execute(query))` but a comprehension is more readable.)

uranusjr commented on airflow/models/dag.py:
```diff
@@ -3113,11 +3136,9 @@ def deactivate_stale_dags(expiration_date, session=NEW_SESSION):
-        ):
+        for dag in session.scalars(
+            select(DagModel).where(DagModel.last_parsed_time < expiration_date, DagModel.is_active)
+        ).all():
```
```suggestion
        ):
```

uranusjr commented on airflow/models/dag.py:
```diff
@@ -3453,12 +3474,11 @@ def get_paused_dag_ids(dag_ids: list[str], session: Session = NEW_SESSION) -> se
+            select(DagModel.dag_id)
+            .where(DagModel.is_paused == expression.true())
+            .where(DagModel.dag_id.in_(dag_ids))
+        ).all()
```
```suggestion
        )
```

I think Mypy would complain here though since `paused_dag_ids` changes type in the next line.

uranusjr commented on airflow/models/dag.py:
```diff
@@ -3518,7 +3541,7 @@ def deactivate_deleted_dags(cls, alive_dag_filelocs: list[str], session=NEW_SESS
 
-        dag_models = session.query(cls).all()
+        dag_models = session.scalars(select(cls)).all()
         for dag_model in dag_models:
```
```suggestion
        for dag_model in session.scalars(select(cls)):
```

(I wonder if this can actually be rewritten with `update`)

ephraimbuddy replied:
Maybe in a different PR

uranusjr commented on airflow/models/dag.py:
```diff
@@ -3539,28 +3562,30 @@ def dags_needing_dagruns(cls, session: Session) -> tuple[Query, dict[str, tuple[
+                .join(DagScheduleDatasetReference.queue_records, isouter=True)
+                .group_by(DagScheduleDatasetReference.dag_id)
+                .having(func.count() == func.sum(case((DDRQ.target_dag_id.is_not(None), 1), else_=0)))
+            ).all()
```
```suggestion
            )
```

uranusjr commented on airflow/models/dag.py:
```diff
@@ -3539,28 +3562,30 @@ def dags_needing_dagruns(cls, session: Session) -> tuple[Query, dict[str, tuple[
+                        .where(DagModel.dag_id.in_(dataset_triggered_dag_ids))
+                        .group_by(DagModel.dag_id)
+                        .having(func.count() >= func.max(DagModel.max_active_runs))
+                    ).all()
```
```suggestion
                    )
```

I think this can be rewritten with `scalars` though

ephraimbuddy committed:
Apply suggestions from code review

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

ephraimbuddy committed:
Use scalars instead of execute and fix typing

uranusjr commented on airflow/models/pool.py:
```diff
@@ -114,7 +114,7 @@ def create_or_update_pool(
             raise ValueError("Pool name must not be empty")
 
-        pool = session.query(Pool).filter_by(pool=name).one_or_none()
+        pool = session.scalars(select(Pool).filter_by(pool=name)).one_or_none()
```
```suggestion
        pool = session.scalar(select(Pool).filter_by(pool=name))
```

for consistency (this is the style we’re using elsewhere)

uranusjr commented on airflow/models/pool.py:
```diff
@@ -162,19 +162,19 @@ def slots_stats(
+            select(TaskInstance.pool, TaskInstance.state, func.sum(TaskInstance.pool_slots))
             .filter(TaskInstance.state.in_(list(EXECUTION_STATES)))
             .group_by(TaskInstance.pool, TaskInstance.state)
         ).all()
```
This `all()` isn’t needed (or rather, it’s not needed previously to begin with)

uranusjr commented on airflow/models/dagrun.py:
```diff
@@ -1335,44 +1335,45 @@ def schedule_tis(
                     )
-                    .update({TI.state: State.SCHEDULED}, synchronize_session=False)
-                )
+                    .values(state=State.SCHEDULED)
```
```suggestion
                    .values(state=TaskInstanceState.SCHEDULED)
```

uranusjr commented on airflow/models/dagrun.py:
```diff
@@ -1335,44 +1335,45 @@ def schedule_tis(
-                            TI.duration: 0,
-                        },
+                    .values(
+                        state=State.SUCCESS,
```
```suggestion
                        state=TaskInstanceState.SUCCESS,
```

uranusjr commented on airflow/models/dag.py:
```diff
@@ -1296,11 +1307,13 @@ def get_concurrency_reached(self, session=NEW_SESSION) -> bool:
+        total_tasks = session.scalar(
+            select(func.count(TI.task_id)).where(
+                TI.dag_id == self.dag_id,
+                TI.state == State.RUNNING,
```
```suggestion
                TI.state == TaskInstanceState.RUNNING,
```

uranusjr commented on airflow/models/dag.py:
```diff
@@ -1402,19 +1415,18 @@ def get_num_active_runs(self, external_trigger=None, only_running=True, session=
+        query = select(func.count()).where(DagRun.dag_id == self.dag_id)
         if only_running:
-            query = query.filter(DagRun.state == State.RUNNING)
+            query = query.where(DagRun.state == State.RUNNING)
```
```suggestion
            query = query.where(DagRun.state == DagRunState.RUNNING)
```

uranusjr commented on airflow/models/dag.py:
```diff
@@ -1402,19 +1415,18 @@ def get_num_active_runs(self, external_trigger=None, only_running=True, session=
+            query = query.where(DagRun.state == State.RUNNING)
         else:
-            query = query.filter(DagRun.state.in_({State.RUNNING, State.QUEUED}))
+            query = query.where(DagRun.state.in_({State.RUNNING, State.QUEUED}))
```
```suggestion
            query = query.where(DagRun.state.in_({DagRunState.RUNNING, DagRunState.QUEUED}))
```

ephraimbuddy committed:
Apply suggestions from code review

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

ephraimbuddy committed:
Apply suggestions from code review

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

ephraimbuddy committed:
remove all

ephraimbuddy committed:
Remove one_or_none where possible
